### PR TITLE
Adding Rake task for creating automatic appointments for defaulters

### PR DESCRIPTION
**Pull request by timmyjose**
Link to story: https://www.pivotaltracker.com/n/projects/2184102/stories/164570825
Task 4: Create a Rake task to create "automatic" appointments for defaulters with the `facility_id` set to the id of facility where the latest BP reading was taken.

### CODE REVIEW

File: lib/tasks/data_migration.rake [hunk 1]
```diff
@@ -79,4 +79,62 @@ namespace :data_migration do
     puts "Associating protocol for #{organization.facility_groups.count} facility groups"
     organization.facility_groups.update(protocol: protocol)
   end
+
+  desc "Create automatic appointment for defaulters so that they show up in the Overdue List"
+  task create_automatic_appointment_for_defaulters: :environment do
+    last_bp_older_than_one_month = ->(p) { p.latest_blood_pressure.blank? || p.latest_blood_pressure.device_created_at < 1.month.ago }
+
+    patient_is_hypertensive = ->(p) { p.latest_blood_pressure.present? && (p.latest_blood_pressure.systolic > 140 || p.latest_blood_pressure.diastolic < 90) }
```
Comment (govindkrjoshi) [inline lib/tasks/data_migration.rake]:
```suggestion
    patient_is_hypertensive = ->(p) { p.latest_blood_pressure.present? && p.latest_blood_pressure.hypertensive? }
```

File: lib/tasks/data_migration.rake [hunk 2]
```diff
@@ -79,4 +79,62 @@ namespace :data_migration do
...
+
+    patient_has_medical_history = ->(p) { (p.medical_history.present? && (p.medical_history.prior_stroke == "yes" || p.medical_history.prior_heart_attack == "yes" || p.medical_history.chronic_kidney_disease == "yes" ||
```
Comment (govindkrjoshi) [inline lib/tasks/data_migration.rake]:
```suggestion
    patient_has_cardio_vascular_history = ->(p) { (p.medical_history.present? && (p.medical_history.prior_stroke == "yes" || p.medical_history.prior_heart_attack == "yes" || p.medical_history.chronic_kidney_disease == "yes" ||
```

Comment (govindkrjoshi) [reply lib/tasks/data_migration.rake]: Use `#indicates_risk` from MedicalHistory

File: lib/tasks/data_migration.rake [hunk 3]
```diff
@@ -79,4 +79,62 @@ namespace :data_migration do
...
+      p.medical_history.diabetes == "yes")) }
+
+    defaulters = Patient.select do |p|
+      p.appointments.count == 0 &&
+        last_bp_older_than_one_month.call(p) &&
+        (patient_is_hypertensive.call(p) || patient_has_medical_history.call(p) || p.prescription_drugs.present?)
```
Comment (govindkrjoshi) [inline lib/tasks/data_migration.rake]: Let's check if this should be true only for protocol drugs.

File: lib/tasks/data_migration.rake [hunk 4]
```diff
@@ -79,4 +79,62 @@ namespace :data_migration do
...
+      app_creation_time = Time.now
+      app_scheduled_date = 32.days.ago # to trigger overdue-ness
+
+      begin
+        automatic_appointment =
+          Appointment.create(patient_id: defaulter.id, facility_id: latest_bp_facility_id,
+                             device_created_at: app_creation_time, device_updated_at: app_creation_time, created_at: app_creation_time, updated_at: app_creation_time,
+                             status: 'scheduled', scheduled_date: app_scheduled_date, appointment_type: Appointment.appointment_types[:automatic])
+
+        automatic_appointment.save
```
Comment (govindkrjoshi) [inline lib/tasks/data_migration.rake]: This is redundant.

File: lib/tasks/data_migration.rake [hunk 5]
```diff
@@ -79,4 +79,62 @@ namespace :data_migration do
...
+        processed_defaulters_count += 1
```
Comment (govindkrjoshi) [inline lib/tasks/data_migration.rake]: We should check if `automatic_appointment.errors.present?` is true and log error messages if present.

File: lib/tasks/data_migration.rake [hunk 6]
```diff
@@ -79,4 +79,62 @@ namespace :data_migration do
...
+    defaulters.each do |defaulter|
+      if defaulter.latest_blood_pressure.blank?
+        puts "Failed to find latest Blood Pressure reading for defaulter #{defaulter.id}. Skipping patient..."
+        unprocessed_or_errored_defaulters_count += 1
+        next
+      end
+
+      latest_bp_facility_id = defaulter.latest_blood_pressure.facility_id
+      app_creation_time = Time.now
+      app_scheduled_date = 32.days.ago # to trigger overdue-ness
```
Comment (govindkrjoshi) [inline lib/tasks/data_migration.rake]: Maybe just do this `40.days.ago` or `patient.latest_blood_pressure.device_created_at + 1.month`?

Comment (govindkrjoshi) [reply lib/tasks/data_migration.rake]: Let's check when to create the appointment.

File: lib/tasks/data_migration.rake (hunk 4 above)
Comment (timmyjose) [reply lib/tasks/data_migration.rake]: Removed `save`.

File: lib/tasks/data_migration.rake (hunk 1 above)
Comment (timmyjose) [reply lib/tasks/data_migration.rake]: Changed to use `hypertensive?`.

File: lib/tasks/data_migration.rake (hunk 2 above)
Comment (timmyjose) [reply lib/tasks/data_migration.rake]: Changed to use `indicates_risk?`

File: lib/tasks/data_migration.rake (hunk 6 above)
Comment (timmyjose) [reply lib/tasks/data_migration.rake]: Changed it to use `40 days ago`.

File: lib/tasks/data_migration.rake (hunk 5 above)
Comment (timmyjose) [reply lib/tasks/data_migration.rake]: Added error handling for the automatic appointment creation.

File: lib/tasks/data_migration.rake [hunk 7]
```diff
@@ -93,4 +93,66 @@ namespace :data_migration do
...
+
+    defaulters.each do |defaulter|
+      if defaulter.latest_blood_pressure.blank?
+        puts "Failed to find latest Blood Pressure reading for defaulter #{defaulter.id}. Skipping patient..."
+        unprocessed_or_errored_defaulters_count += 1
+        next
+      end
+
+      latest_bp_facility_id = defaulter.latest_blood_pressure.facility_id
+      app_creation_time = Time.now
```
Comment (govindkrjoshi) [inline lib/tasks/data_migration.rake]: Use latest blood pressures `device_created_at` and `app_creation_time + 30.days` for `app_creation_time` and `app_scheduled_date`

**Pull request merged**